doc/glossary: host, build and target systems (#8570)
NixOS/nix: doc/manual/src/glossary.md

    See [installables](./command-ref/new-cli/nix.md#installables) for [`nix` commands](./command-ref/new-cli/nix.md) (experimental) for details.

  - [platform]{#gloss-platform}

    [platform]: #gloss-platform

>     A set of execution environments capable of executing some set of software. A platform is typically described by attributes such as instruction set architecture, operating system, and application binary interface.

Reviewer: "What does "a set of" mean?
```suggestion
    An execution environments capable of executing some set of software. A platform is typically described by attributes such as instruction set architecture, operating system, and application binary interface.
```
It reads fine to me without it

>     A set of execution environments capable of executing some set of software. A platform is typically described by attributes such as instruction set architecture, operating system, and application binary interface.

Reviewer 2: I suppose this is meant to cover emulating environments as well. I think it's correct but maybe too clever.

>     A set of execution environments capable of executing some set of software. A platform is typically described by attributes such as instruction set architecture, operating system, and application binary interface.

Reviewer: Hmm I still wouldn't use set then. I don't think there is an *objective* basis with which to say e.g. "linux and windows syscalls via winelib is 2 interfaces, linux syscalls is one interface". They both are interfaces, it is just interfaces form partial orders.

(Sets are a way to get partial orders out of discrete/unordered things, but I am saying we never had discrete/unordered things to begin with. The concept of sets is not needed.)

    Nix is not intimately aware of all possible platforms.
    Rather, it abstracts the platform into

>      - [system type]{#gloss-system-type} string

Reviewer 2: The system type definition lacks an anchor

```suggestion
     - [system type]string
```

     - [system features](command-ref/conf-file.md#system-features), to a lesser extent, as these are often thought of as additions to a platform, rather being part of the platform definition.

  - [system type]:{#gloss-system-type}

>     An identifier for a specific [platform] or (in rare cases) a class of compatible platforms, such as `x86_64-v3-linux`.

Reviewer 2: ```suggestion
    An identifier for a specific [platform] or (in rare cases) a class of compatible platforms, such as `x86_64-linux`.
```

This seems to be atypical for Nix. Why not something everyone knows? Wouldn't be as surprising.

    To build a derivation locally, the [derivation `system` field](language/derivations.md#attr-system) must occur in the local Nix installation's [`system`](command-ref/conf-file.md#conf-system) or [`extra-platforms`](command-ref/conf-file.md#conf-extra-platforms) settings, among other rules.
    Otherwise, a [remote build](advanced-topics/distributed-builds.md) is needed.

  - [build platform]{#gloss-build-platform}

    [build platform]: #gloss-build-platform

>     The [platform] that performs the build or compilation, typically a dash-separated string.

Reviewer 2: ```suggestion
    The [platform] that performs the build or compilation, typically denoted by a [system type] string.
```

Same for the other occurrences.

Author: ```suggestion
    The [platform] that performs the build or compilation, typically identified by a [system type] string.
```

Denote would suggest that there's a syntax to it.

**denotation** (computer science) Any mathematical object which describes the meanings of expressions from the languages, formalized in the theory of denotational semantics.

While the system string does have somewhat of a language in practice, it's something Nix does not need to know about, as mentioned elsewhere.

>     A program that is produced by a [derivation] has a [build configuration] with a [build platform] that can be executed by the [derivation system](language/derivations.md#attr-system).
>     After it has been built, the [build platform] is no longer relevant.

Reviewer 2: No need to link the term to itself. All the links are already a strain to read.

As discussed today:

```suggestion
    A program that is produced by a [derivation] must have a [build configuration] where the build platform matches the [`system` field of that derivation](language/derivations.md#attr-system).
    After it has been built, the build platform is no longer relevant.
```

Same for the other occurrences.

    This term is based on the [GCC system terminology](#gloss-gcc-system-terminology).

  - [host platform]{#gloss-host-platform}

    [host platform]: #gloss-host-platform

    The [platform] where build products will run on, typically a dash-separated string.

>     Programs that run natively in a [derivation] must have a [build configuration] with a [host platform] that can by executed by the [derivation system](language/derivations.md#attr-system).

Reviewer 2: ```suggestion
    Programs that run when a [derivation] is built must have <as before>
```

Not sure here. Apparently you're using "in a derivation" with the meaning "while running the `builder` executable of a derivation`, which is often abbreviated by "when building a derivation", and I don't think it's semantically wrong, but given that *derivation* is such a mystical term for beginners, we should be careful with condensing too much. I'd rather cut down on "building" a derivation or at least define what that *actually* means in the glossary as well. Actually this is something that I wanted to do some time ago but got rabbit-holed with.

Reviewer: I think while the build platform is only a property of the build configuration, the host and target platforms are properties of the built program itself too. I think it makes sense to break symmetry and stop talking about the build configuration for these latter too to further emphasize what information is relevant when.

    This term is based on the [GCC system terminology](#gloss-gcc-system-terminology).

  - [target platform]{#gloss-target-platform}

    [target platform]: #gloss-target-platform

    The [platform] for which a compiler will generate code, typically a dash-separated string.

>     When discussing a build that does not produce a compiler, we only use [build platform] and [host platform] to describe it. This also applies to compilers with a dynamic target, which is not part of the [build configuration].

Reviewer 2: This should be rephrased, highlighting that it only applies to a few special cases where the target platform is baked into the compiler binary, such as GCC (link to where it matters).

The way it's currently phrased implies it's the most natural thing, and somewhat unusual to have a "dynamic target". Which I think is absolutely not intuitive if you haven't been digging in the bowels of obscure C code in the past 40 years.

>     We try to avoid discussing the actual build from the perspective of its compiler, instead focusing on the build itself.
>     Otherwise we won't have a term to distinguish the actual [target platform] from what we should have called the [host platform].

Reviewer 2: ```suggestion
```

I don't understand what this is trying to tell me.

Reviewer: A lot of compiler writers screwed this up, because they only thought of the terms from *their* perspective and not *their user's* perspective.

>     If the output of a compiler is to be run in a [derivation], the compiler's [target platform] must be executable by the [derivation system](language/derivations.md#attr-system).

Reviewer 2: I think here the "match" terminology would be especially helpful.

    This term is based on the [GCC system terminology](#gloss-gcc-system-terminology).

  - [GCC system terminology]{#gloss-gcc-system-terminology}

    The Nix ecosystem uses the [GCC system terminology](https://gcc.gnu.org/onlinedocs/gccint/Configure-Terms.html) for the names and roles of _build_, _host_, and _target_, but may apply the term _cross_ more loosely.

>     See [host platform], [build platform] and [target platform] for how each relates to [derivation].

Reviewer 2: ```suggestion
    See [host platform], [build platform] and [target platform] for how each relates to a [derivation].
```

  - [build configuration]{#gloss-build-configuration}

    [build configuration]: #gloss-build-configuration

    The software configuration whose effects enable the build process and most of which also influence the [build product](#gloss-build-product).

    Build configuration does not have a single agreed upon definition, but consists of a potentially broad set of settings that include:

>     - [build platform], often implicitly

Reviewer 2: ```suggestion
    - [build platform], often implicitly meaning the current system
```

    - [host platform]

>     - [target platform], only when building a compiler with a static target platform

Reviewer 2: ```suggestion
    - [target platform], only for compilers where their compilation target platform is determined when the compiler itself is build
```

This is a bit lengthy and clumsy, but  think we should make that unambiguous. Maybe you have a more elegant way of saying this without obscuring the meaning.

    - options passed to a compiler, such optimization settings and the host platform (when using a compiler with a dynamic target platform)
    - dependency options, such as which build tools and dependencies to use
    - anything specified in a custom build script

>     In the context of Nix, the [build configuration] can be thought of as a [derivation] minus the source code, plus intentional build impurities.

Reviewer 2: ```suggestion
```

This seems misleading. Which source code? In reality, the build configuration is part of the source code. I'd rather leave that out.

We want to help build simplified mental models that don't have to be revised substantially as one becomes more expert.

Reviewer: The build configuration is a context for *something*. I am having a hard time finding a reference for this (https://ncatlab.org/nlab/show/context lol) but contexts are sort of dual to expressions --- not the nice self-contained thing, but the expansive thing with no clear external boundary but a nice clear slot in the middle.

It sounds to me what robert is saying "take the thing we are trying to build" out of the donut hole, now what are we left with? That sounds like a good intuition for me. Even if the Nix code is first-party, not 3rd party packaging like Nixpkgs, we can stil distinguish the parts of hte source code that evaluate to derivations from the parts that get copied to the store and referenced from within derivations.

>   - [build product]{#gloss-build-product}

Reviewer 2: ```suggestion
  - [build result]{#gloss-build-product}
```

We started calling it that in the architecture docs with @Ericson2314 and it's already a term in the manual and nix.dev. It doesn't meen it have to be, but I like it because it doesn't raise mathematical associations.

>     The result of running a build process. When a build process runs in a [derivation], build products are typically copied into the [output path](#gloss-output-path)(s).

Reviewer 2: ```suggestion
    The result of running a build process. When a `builder` executable of a [derivation] is run, build products are typically copied into the [output path](#gloss-output-path)(s).
```

I don't like this yet though. It ties into defining what "building" means. We use that term very liberally and just assume people know what it means precisely. But it has a precise definition that's just not stated anywhere.

    See [installables](./command-ref/new-cli/nix.md#installables) for [`nix` commands](./command-ref/new-cli/nix.md) (experimental) for details.

  - [build system]{#gloss-build-system}

    [build system]: #gloss-build-system

    A representation of the processor architecture and platform that performs the build or compilation, typically a dash-separated string.

    This is based on the [GCC system terminology](#gloss-gcc-system-terminology).

>     Programs that are produced by a [derivation] were configured to have a [build system] that can be executed by the [derivation system](language/derivations.md#attr-system). When using the program, only its [host system] is still relevant.

Reviewer: ```suggestion
    Programs that are produced by a [derivation] were configured to have a [build system] that can be executed by the [derivation system](language/derivations.md#attr-system).

    Conversely, when using the program after it is been built, the [build system] is no longer relevant.
```
Both host and target (if it exists) are relevant after building.

  - [host system]{#gloss-host-system}

    [host system]: #gloss-host-system

    A representation of the processor architecture and platform where build products will run on, typically a dash-separated string.

    This is based on the [GCC system terminology](#gloss-gcc-system-terminology).

    Programs that run natively in a [derivation] must be configured with a [host system] that can by executed by the [derivation system](language/derivations.md#attr-system).

  - [target system]{#gloss-target-system}

    [target system]: #gloss-target-system

    A representation of the processor architecture and platform that a compiler will produce, typically a dash-separated string.

>     When discussing a build that does not produce a compiler, we only use [build system] and [host system] to describe it; we try to avoid discussing the actual build from the perspective its compiler's configuration.

Reviewer: ```suggestion
    When discussing a build that does not produce a compiler, we only use [build system] and [host system] to describe it.

    We try to avoid discussing the actual build from the perspective its compiler's configuration, and instead always talk about the build from the perspective of the build itself.
    Avoiding that ensures that if we are building a compiler do use "[target system]", it is clear we are referring to the target of the compiler being built, and not the target of a preexisting compiler used during the build.
```
I don't know if that is too much a digression for Nix, but I do like to make the ambiguity we're avoiding explicit, rather than left to the reader's imagination :).

Author: ```suggestion
    When discussing a build that does not produce a compiler, we only use [build system] and [host system] to describe it.
    
    We try to avoid discussing the actual build from the perspective of its compiler, instead focusing on the build itself.
    Otherwise you won't have a term to distinguish the actual [target system] from what you should have called the [host system].
```

I think I've managed to simplify it. Not sure if I should have tried this in the morning :sweat_smile:

Author: @Ericson2314 wdyt?

    See [installables](./command-ref/new-cli/nix.md#installables) for [`nix` commands](./command-ref/new-cli/nix.md) (experimental) for details.

  - [build system]{#gloss-build-system}

    [build system]: #gloss-build-system

    A representation of the processor architecture and platform that performs the build or compilation, typically a dash-separated string.

    This is based on the [GCC system terminology](#gloss-gcc-system-terminology).

>     Programs that are produced by a [derivation] were configured when they were being built to have a [build system] that can be executed by the [derivation system](language/derivations.md#attr-system).
>     Conversely, when using the program after it is been built, the [build system] is no longer relevant.

Author: ```suggestion
    A program that is produced by a [derivation] is built with a [build system] that can be executed by the [derivation system](language/derivations.md#attr-system).
    After it has been built, the [build system] is no longer relevant.
```

Simplifying a bit and I don't think it was a particularly important "converse" relationship.
